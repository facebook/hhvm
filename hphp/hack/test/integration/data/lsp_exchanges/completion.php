<?hh  //strict

function testing_area(): void {

}

/** test_function docblock. */
function test_function(): void {

}

interface CompletionInterface {
  /** Doc block should fall back to interface. */
  public function interfaceDocBlockMethod(): void;
}

class CompletionClass {
  public function interfaceDocBlockMethod(): void {}
}
