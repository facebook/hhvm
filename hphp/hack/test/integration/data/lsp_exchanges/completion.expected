[
  {
    "jsonrpc": "2.0",
    "id": 1,
    "result": {
      "capabilities": {
        "textDocumentSync": {
          "openClose": true,
          "change": 2,
          "willSave": false,
          "willSaveWaitUntil": false,
          "save": {
            "includeText": false
          }
        },
        "hoverProvider": true,
        "completionProvider": {
          "resolveProvider": true,
          "triggerCharacters": [
            "$",
            ">",
            "\\",
            ":",
            "<",
            "[",
            "'",
            "\""
          ]
        },
        "signatureHelpProvider": {
          "triggerCharacters": [
            "(",
            ","
          ]
        },
        "definitionProvider": true,
        "referencesProvider": true,
        "documentHighlightProvider": true,
        "documentSymbolProvider": true,
        "workspaceSymbolProvider": true,
        "codeActionProvider": false,
        "documentFormattingProvider": true,
        "documentRangeFormattingProvider": true,
        "documentOnTypeFormattingProvider": {
          "firstTriggerCharacter": ";",
          "moreTriggerCharacter": [
            "}"
          ]
        },
        "renameProvider": true,
        "typeCoverageProvider": true,
        "rageProvider": true
      }
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 20,
    "result": {
      "isIncomplete": false,
      "items": [
        {
          "label": "ab:cd:alpha",
          "kind": 7,
          "detail": "class",
          "inlineDetail": "class",
          "insertTextFormat": 1,
          "textEdit": {
            "range": {
              "start": {
                "line": 3,
                "character": 6
              },
              "end": {
                "line": 3,
                "character": 6
              }
            },
            "newText": "ab:cd:alpha"
          },
          "data": {
            "filename": "${root_path}${path_sep}completion_extras.php",
            "line": 9,
            "char": 13,
            "base_class": null
          }
        },
        {
          "label": "ab:cd:text",
          "kind": 7,
          "detail": "class",
          "inlineDetail": "class",
          "insertTextFormat": 1,
          "textEdit": {
            "range": {
              "start": {
                "line": 3,
                "character": 6
              },
              "end": {
                "line": 3,
                "character": 6
              }
            },
            "newText": "ab:cd:text"
          },
          "data": {
            "filename": "${root_path}${path_sep}completion_extras.php",
            "line": 4,
            "char": 13,
            "base_class": null
          }
        }
      ]
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 30,
    "result": {
      "isIncomplete": false,
      "items": [
        {
          "label": "ab:cd:alpha",
          "kind": 7,
          "detail": "class",
          "inlineDetail": "class",
          "insertTextFormat": 1,
          "textEdit": {
            "range": {
              "start": {
                "line": 3,
                "character": 6
              },
              "end": {
                "line": 3,
                "character": 7
              }
            },
            "newText": "ab:cd:alpha"
          },
          "data": {
            "filename": "${root_path}${path_sep}completion_extras.php",
            "line": 9,
            "char": 13,
            "base_class": null
          }
        },
        {
          "label": "ab:cd:text",
          "kind": 7,
          "detail": "class",
          "inlineDetail": "class",
          "insertTextFormat": 1,
          "textEdit": {
            "range": {
              "start": {
                "line": 3,
                "character": 6
              },
              "end": {
                "line": 3,
                "character": 7
              }
            },
            "newText": "ab:cd:text"
          },
          "data": {
            "filename": "${root_path}${path_sep}completion_extras.php",
            "line": 4,
            "char": 13,
            "base_class": null
          }
        }
      ]
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 32,
    "result": {
      "isIncomplete": false,
      "items": [
        {
          "label": "ab:cd:alpha",
          "kind": 7,
          "detail": "class",
          "inlineDetail": "class",
          "insertTextFormat": 1,
          "textEdit": {
            "range": {
              "start": {
                "line": 3,
                "character": 6
              },
              "end": {
                "line": 3,
                "character": 9
              }
            },
            "newText": "ab:cd:alpha"
          },
          "data": {
            "filename": "${root_path}${path_sep}completion_extras.php",
            "line": 9,
            "char": 13,
            "base_class": null
          }
        },
        {
          "label": "ab:cd:text",
          "kind": 7,
          "detail": "class",
          "inlineDetail": "class",
          "insertTextFormat": 1,
          "textEdit": {
            "range": {
              "start": {
                "line": 3,
                "character": 6
              },
              "end": {
                "line": 3,
                "character": 9
              }
            },
            "newText": "ab:cd:text"
          },
          "data": {
            "filename": "${root_path}${path_sep}completion_extras.php",
            "line": 4,
            "char": 13,
            "base_class": null
          }
        }
      ]
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 40,
    "result": {
      "isIncomplete": false,
      "items": [
        {
          "label": "width",
          "kind": 10,
          "detail": "?int",
          "inlineDetail": "?int",
          "insertTextFormat": 1,
          "textEdit": {
            "range": {
              "start": {
                "line": 3,
                "character": 17
              },
              "end": {
                "line": 3,
                "character": 17
              }
            },
            "newText": "width"
          },
          "data": {
            "filename": "${root_path}${path_sep}completion_extras.php",
            "line": 5,
            "char": 27,
            "base_class": "\\:ab:cd:text"
          }
        },
        {
          "label": "color",
          "kind": 10,
          "detail": "?string",
          "inlineDetail": "?string",
          "insertTextFormat": 1,
          "textEdit": {
            "range": {
              "start": {
                "line": 3,
                "character": 17
              },
              "end": {
                "line": 3,
                "character": 17
              }
            },
            "newText": "color"
          },
          "data": {
            "filename": "${root_path}${path_sep}completion_extras.php",
            "line": 5,
            "char": 13,
            "base_class": "\\:ab:cd:text"
          }
        }
      ]
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 50,
    "result": {
      "isIncomplete": false,
      "items": [
        {
          "label": "width",
          "kind": 10,
          "detail": "?int",
          "inlineDetail": "?int",
          "insertTextFormat": 1,
          "textEdit": {
            "range": {
              "start": {
                "line": 3,
                "character": 17
              },
              "end": {
                "line": 3,
                "character": 18
              }
            },
            "newText": "width"
          },
          "data": {
            "filename": "${root_path}${path_sep}completion_extras.php",
            "line": 5,
            "char": 27,
            "base_class": "\\:ab:cd:text"
          }
        },
        {
          "label": "color",
          "kind": 10,
          "detail": "?string",
          "inlineDetail": "?string",
          "insertTextFormat": 1,
          "textEdit": {
            "range": {
              "start": {
                "line": 3,
                "character": 17
              },
              "end": {
                "line": 3,
                "character": 18
              }
            },
            "newText": "color"
          },
          "data": {
            "filename": "${root_path}${path_sep}completion_extras.php",
            "line": 5,
            "char": 13,
            "base_class": "\\:ab:cd:text"
          }
        }
      ]
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 60,
    "result": {
      "isIncomplete": false,
      "items": [
        {
          "label": ":ab:cd:alpha",
          "kind": 4,
          "detail": "function(): :ab:cd:alpha",
          "inlineDetail": "()",
          "itemType": ":ab:cd:alpha",
          "insertTextFormat": 1,
          "textEdit": {
            "range": {
              "start": {
                "line": 3,
                "character": 9
              },
              "end": {
                "line": 3,
                "character": 10
              }
            },
            "newText": ":ab:cd:alpha"
          },
          "data": {
            "filename": "${root_path}${path_sep}completion_extras.php",
            "line": 9,
            "char": 13,
            "base_class": "\\:ab:cd:alpha"
          }
        },
        {
          "label": ":ab:cd:text",
          "kind": 4,
          "detail": "function(): :ab:cd:text",
          "inlineDetail": "()",
          "itemType": ":ab:cd:text",
          "insertTextFormat": 1,
          "textEdit": {
            "range": {
              "start": {
                "line": 3,
                "character": 9
              },
              "end": {
                "line": 3,
                "character": 10
              }
            },
            "newText": ":ab:cd:text"
          },
          "data": {
            "filename": "${root_path}${path_sep}completion_extras.php",
            "line": 4,
            "char": 13,
            "base_class": "\\:ab:cd:text"
          }
        }
      ]
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 70,
    "result": {
      "isIncomplete": false,
      "items": [
        {
          "label": ":ab:cd:alpha",
          "kind": 4,
          "detail": "function(): :ab:cd:alpha",
          "inlineDetail": "()",
          "itemType": ":ab:cd:alpha",
          "insertTextFormat": 1,
          "textEdit": {
            "range": {
              "start": {
                "line": 3,
                "character": 9
              },
              "end": {
                "line": 3,
                "character": 11
              }
            },
            "newText": ":ab:cd:alpha"
          },
          "data": {
            "filename": "${root_path}${path_sep}completion_extras.php",
            "line": 9,
            "char": 13,
            "base_class": "\\:ab:cd:alpha"
          }
        },
        {
          "label": ":ab:cd:text",
          "kind": 4,
          "detail": "function(): :ab:cd:text",
          "inlineDetail": "()",
          "itemType": ":ab:cd:text",
          "insertTextFormat": 1,
          "textEdit": {
            "range": {
              "start": {
                "line": 3,
                "character": 9
              },
              "end": {
                "line": 3,
                "character": 11
              }
            },
            "newText": ":ab:cd:text"
          },
          "data": {
            "filename": "${root_path}${path_sep}completion_extras.php",
            "line": 4,
            "char": 13,
            "base_class": "\\:ab:cd:text"
          }
        }
      ]
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 80,
    "result": {
      "label": ":ab:cd:alpha",
      "kind": 4,
      "detail": "function(): :ab:cd:alpha",
      "inlineDetail": "()",
      "itemType": ":ab:cd:alpha",
      "documentation": ":ab:cd:alpha docblock",
      "insertText": ":ab:cd:alpha",
      "insertTextFormat": 1,
      "data": {
        "filename": "${root_path}${path_sep}completion_extras.php",
        "line": 9,
        "char": 13
      }
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 90,
    "result": {
      "isIncomplete": false,
      "items": [
        {
          "label": ":width",
          "kind": 10,
          "detail": "?int",
          "inlineDetail": "?int",
          "insertTextFormat": 1,
          "textEdit": {
            "range": {
              "start": {
                "line": 3,
                "character": 29
              },
              "end": {
                "line": 3,
                "character": 29
              }
            },
            "newText": ":width"
          },
          "data": {
            "filename": "${root_path}${path_sep}completion_extras.php",
            "line": 5,
            "char": 27,
            "base_class": "\\:ab:cd:text"
          }
        },
        {
          "label": ":color",
          "kind": 10,
          "detail": "?string",
          "inlineDetail": "?string",
          "insertTextFormat": 1,
          "textEdit": {
            "range": {
              "start": {
                "line": 3,
                "character": 29
              },
              "end": {
                "line": 3,
                "character": 29
              }
            },
            "newText": ":color"
          },
          "data": {
            "filename": "${root_path}${path_sep}completion_extras.php",
            "line": 5,
            "char": 13,
            "base_class": "\\:ab:cd:text"
          }
        }
      ]
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 100,
    "result": {
      "isIncomplete": false,
      "items": [
        {
          "label": ":width",
          "kind": 10,
          "detail": "?int",
          "inlineDetail": "?int",
          "insertTextFormat": 1,
          "textEdit": {
            "range": {
              "start": {
                "line": 3,
                "character": 29
              },
              "end": {
                "line": 3,
                "character": 30
              }
            },
            "newText": ":width"
          },
          "data": {
            "filename": "${root_path}${path_sep}completion_extras.php",
            "line": 5,
            "char": 27,
            "base_class": "\\:ab:cd:text"
          }
        },
        {
          "label": ":color",
          "kind": 10,
          "detail": "?string",
          "inlineDetail": "?string",
          "insertTextFormat": 1,
          "textEdit": {
            "range": {
              "start": {
                "line": 3,
                "character": 29
              },
              "end": {
                "line": 3,
                "character": 30
              }
            },
            "newText": ":color"
          },
          "data": {
            "filename": "${root_path}${path_sep}completion_extras.php",
            "line": 5,
            "char": 13,
            "base_class": "\\:ab:cd:text"
          }
        }
      ]
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 110,
    "result": {
      "isIncomplete": false,
      "items": [
        {
          "label": "test_function",
          "kind": 3,
          "detail": "function(): void",
          "inlineDetail": "()",
          "itemType": "void",
          "insertTextFormat": 1,
          "textEdit": {
            "range": {
              "start": {
                "line": 3,
                "character": 0
              },
              "end": {
                "line": 3,
                "character": 8
              }
            },
            "newText": "test_function"
          },
          "data": {
            "filename": "${root_path}${path_sep}completion.php",
            "line": 8,
            "char": 10,
            "base_class": null
          }
        }
      ]
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 120,
    "result": {
      "label": "test_function",
      "kind": 3,
      "detail": "function(): void",
      "inlineDetail": "()",
      "itemType": "void",
      "documentation": "test_function docblock.",
      "insertText": "test_function",
      "insertTextFormat": 1,
      "data": {
        "filename": "${root_path}${path_sep}completion.php",
        "line": 8,
        "char": 10
      }
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 130,
    "result": {
      "isIncomplete": false,
      "items": []
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 140,
    "result": {
      "isIncomplete": false,
      "items": [
        {
          "label": "class",
          "kind": 12,
          "detail": "classname<this>",
          "inlineDetail": "classname<this>",
          "insertTextFormat": 1,
          "textEdit": {
            "range": {
              "start": {
                "line": 3,
                "character": 35
              },
              "end": {
                "line": 3,
                "character": 35
              }
            },
            "newText": "class"
          },
          "data": {
            "filename": "${root_path}${path_sep}completion_extras.php",
            "line": 13,
            "char": 6,
            "base_class": "\\Elsa"
          }
        },
        {
          "label": "Bard",
          "kind": 12,
          "detail": "Elsa",
          "inlineDetail": "Elsa",
          "insertTextFormat": 1,
          "textEdit": {
            "range": {
              "start": {
                "line": 3,
                "character": 35
              },
              "end": {
                "line": 3,
                "character": 35
              }
            },
            "newText": "Bard"
          },
          "data": {
            "filename": "${root_path}${path_sep}completion_extras.php",
            "line": 13,
            "char": 12,
            "base_class": "\\Elsa"
          }
        },
        {
          "label": "Alonso",
          "kind": 12,
          "detail": "Elsa",
          "inlineDetail": "Elsa",
          "insertTextFormat": 1,
          "textEdit": {
            "range": {
              "start": {
                "line": 3,
                "character": 35
              },
              "end": {
                "line": 3,
                "character": 35
              }
            },
            "newText": "Alonso"
          },
          "data": {
            "filename": "${root_path}${path_sep}completion_extras.php",
            "line": 13,
            "char": 12,
            "base_class": "\\Elsa"
          }
        },
        {
          "label": "isValid",
          "kind": 2,
          "detail": "function(mixed $value): bool",
          "inlineDetail": "(mixed $value)",
          "itemType": "bool",
          "insertTextFormat": 1,
          "textEdit": {
            "range": {
              "start": {
                "line": 3,
                "character": 35
              },
              "end": {
                "line": 3,
                "character": 35
              }
            },
            "newText": "isValid"
          },
          "data": {
            "filename": "${hhi_path}${path_sep}BuiltinEnum.hhi",
            "line": 49,
            "char": 32,
            "base_class": "\\Elsa"
          }
        },
        {
          "label": "getValues",
          "kind": 2,
          "detail": "function(): darray<string, Elsa>",
          "inlineDetail": "()",
          "itemType": "darray<string, Elsa>",
          "insertTextFormat": 1,
          "textEdit": {
            "range": {
              "start": {
                "line": 3,
                "character": 35
              },
              "end": {
                "line": 3,
                "character": 35
              }
            },
            "newText": "getValues"
          },
          "data": {
            "filename": "${hhi_path}${path_sep}BuiltinEnum.hhi",
            "line": 34,
            "char": 32,
            "base_class": "\\Elsa"
          }
        },
        {
          "label": "getNames",
          "kind": 2,
          "detail": "function(): darray<Elsa, string>",
          "inlineDetail": "()",
          "itemType": "darray<Elsa, string>",
          "insertTextFormat": 1,
          "textEdit": {
            "range": {
              "start": {
                "line": 3,
                "character": 35
              },
              "end": {
                "line": 3,
                "character": 35
              }
            },
            "newText": "getNames"
          },
          "data": {
            "filename": "${hhi_path}${path_sep}BuiltinEnum.hhi",
            "line": 43,
            "char": 32,
            "base_class": "\\Elsa"
          }
        },
        {
          "label": "coerce",
          "kind": 2,
          "detail": "function(mixed $value): ?Elsa",
          "inlineDetail": "(mixed $value)",
          "itemType": "?Elsa",
          "insertTextFormat": 1,
          "textEdit": {
            "range": {
              "start": {
                "line": 3,
                "character": 35
              },
              "end": {
                "line": 3,
                "character": 35
              }
            },
            "newText": "coerce"
          },
          "data": {
            "filename": "${hhi_path}${path_sep}BuiltinEnum.hhi",
            "line": 56,
            "char": 32,
            "base_class": "\\Elsa"
          }
        },
        {
          "label": "assertAll",
          "kind": 2,
          "detail": "function(Traversable<mixed> $values): Container<Elsa>",
          "inlineDetail": "(Traversable<mixed> $values)",
          "itemType": "Container<Elsa>",
          "insertTextFormat": 1,
          "textEdit": {
            "range": {
              "start": {
                "line": 3,
                "character": 35
              },
              "end": {
                "line": 3,
                "character": 35
              }
            },
            "newText": "assertAll"
          },
          "data": {
            "filename": "${hhi_path}${path_sep}BuiltinEnum.hhi",
            "line": 70,
            "char": 32,
            "base_class": "\\Elsa"
          }
        },
        {
          "label": "assert",
          "kind": 2,
          "detail": "function(mixed $value): Elsa",
          "inlineDetail": "(mixed $value)",
          "itemType": "Elsa",
          "insertTextFormat": 1,
          "textEdit": {
            "range": {
              "start": {
                "line": 3,
                "character": 35
              },
              "end": {
                "line": 3,
                "character": 35
              }
            },
            "newText": "assert"
          },
          "data": {
            "filename": "${hhi_path}${path_sep}BuiltinEnum.hhi",
            "line": 63,
            "char": 32,
            "base_class": "\\Elsa"
          }
        }
      ]
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 145,
    "result": {
      "label": "isValid",
      "kind": 2,
      "detail": "function(mixed $value): bool",
      "inlineDetail": "(mixed $value)",
      "itemType": "bool",
      "documentation": "\nReturns whether or not the value is defined as a constant.\n",
      "insertTextFormat": 1,
      "textEdit": {
        "range": {
          "start": {
            "line": 3,
            "character": 35
          },
          "end": {
            "line": 3,
            "character": 35
          }
        },
        "newText": "isValid"
      },
      "data": {
        "filename": "${hhi_path}${path_sep}BuiltinEnum.hhi",
        "line": 49,
        "char": 32
      }
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 150,
    "result": {
      "isIncomplete": false,
      "items": []
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 160,
    "result": {
      "isIncomplete": false,
      "items": [
        {
          "label": "test_func",
          "kind": 3,
          "detail": "function(): void",
          "inlineDetail": "()",
          "itemType": "void",
          "insertTextFormat": 1,
          "textEdit": {
            "range": {
              "start": {
                "line": 3,
                "character": 7
              },
              "end": {
                "line": 3,
                "character": 7
              }
            },
            "newText": "test_func"
          },
          "data": {
            "filename": "${root_path}${path_sep}completion_extras_namespace.php",
            "line": 5,
            "char": 10,
            "base_class": null
          }
        }
      ]
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 170,
    "result": {
      "isIncomplete": false,
      "items": [
        {
          "label": "interfaceDocBlockMethod",
          "kind": 2,
          "detail": "function(): void",
          "inlineDetail": "()",
          "itemType": "void",
          "insertTextFormat": 1,
          "textEdit": {
            "range": {
              "start": {
                "line": 3,
                "character": 34
              },
              "end": {
                "line": 3,
                "character": 41
              }
            },
            "newText": "interfaceDocBlockMethod"
          },
          "data": {
            "filename": "${root_path}${path_sep}completion.php",
            "line": 18,
            "char": 19,
            "base_class": "\\CompletionClass"
          }
        }
      ]
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 180,
    "result": {
      "label": "interfaceDocBlockMethod",
      "kind": 2,
      "detail": "function(): void",
      "inlineDetail": "()",
      "itemType": "void",
      "insertTextFormat": 1,
      "textEdit": {
        "range": {
          "start": {
            "line": 3,
            "character": 34
          },
          "end": {
            "line": 3,
            "character": 41
          }
        },
        "newText": "interfaceDocBlockMethod"
      },
      "data": {
        "filename": "${root_path}${path_sep}completion.php",
        "line": 18,
        "char": 19
      }
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 999,
    "result": null
  }
]
